# Activity 3
## Creating a simple model

In this activity we will create a simple logistic regression model from the scikit learn package.
We will then create some model evaluation metrics and test the predictions against those model evaluation metrics.
Let's load the feature data from the first excerice and the target data from the second activity, the feature data from the second activity can also be used.

We should always approach training any machine learning model training as an iterative approach, beginning first with a simple model, and using model evaluation metrics to evaluate the performance of the models.

In [12]:
import pandas as pd
feats = pd.read_csv('data/bank_data_feats_e1.csv', index_col=0)
target = pd.read_csv('data/bank_data_target_a2.csv', index_col=0)

Lets import the logistic regression model and evaluation metrics from scikit learn

In [19]:
from sklearn.linear_model import LogisticRegression
from sklearn import metrics
from sklearn.model_selection import train_test_split

We first begin by creating a test and train dataset. We will train the data using the training dataset and evaluate the performance of the model on the test dataset. Later in the lesson we will add validation datasets that will help us tune the hyperparameters.

We will use a test_size = 0.2 which means that 20% of the data will be reserved for testing

In [16]:
test_size = 0.2
random_state = 42
X_train, X_test, y_train, y_test = train_test_split(feats, target, test_size=test_size, random_state=random_state)

Let's make sure our dimensions are correct

In [17]:
print(f'Shape of X_train: {X_train.shape}')
print(f'Shape of y_train: {y_train.shape}')
print(f'Shape of X_test: {X_test.shape}')
print(f'Shape of y_test: {y_test.shape}')

Shape of X_train: (3616, 32)
Shape of y_train: (3616, 1)
Shape of X_test: (905, 32)
Shape of y_test: (905, 1)


We fit our model first by instantiating it, then by fitting the model to the training data

In [21]:
model = LogisticRegression()
model.fit(X_train, y_train['y'])

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

To test the model performance we will predict the outcome on the test features (X_test), and compare those outcomes to real values (y_test)

In [23]:
y_pred = model.predict(X_test)

Now let's compare against the true values. Let's start by using accuracy, accuracy is defined as the propotion of correct predictions out of the total predictions.

In [25]:
accuracy = metrics.accuracy_score(y_pred=y_pred, y_true=y_test)
print(f'Accuracy of the model is {accuracy*100:.4f}%')

Accuracy of the model is 89.9448%


89.9448% - that's not bad with for a simple model with little feature engineering!

### Other evaluation metrics

Other common metrics in classification models are precision, recall, and f1-score.
Recall is defined as the proportion of correct positive predictions relative to total true postive values. Precision is defined as the proportion of correct positive predictions relative to total predicted postive values. F1 score is a combination of precision and recall, defined as 2 times the product of precision and recall, divided by the sum of the two.

It's useful to use these other evaluation metrics other than accuracy when the distribution of true and false values. We want these values to be as close to 1.0 as possible.

In [43]:
precision, recall, fscore, _ = metrics.precision_recall_fscore_support(y_pred=y_pred, y_true=y_test, average='binary')
print(f'Precision: {precision:.4f}\nRecall: {recall:.4f}\nfscore: {fscore:.4f}')

Precision: 0.5814
Recall: 0.2551
fscore: 0.3546


We can see here that while the accuracy is high, the recall is low, which means that we're missing most of the true positive values.

### Feature importances
   

We can look at which features are important by looking at the magnitude of the coefficients. Those with a larger coefficients will have a greater contribution to the result. Those with a positive value will make the result head toward the true result, that the customer will not subscribe. Features with a negative value for the coefficient will make the result heads towards a false result, that the customer will not subscribe to the product.

As a note, since the features were not normalized (having the same scale), the values for these coefficients shouls serve as a rough guide as to observe which features add predictive power.

In [37]:
coef_list = [f'{feature}: {coef}' for coef, feature in sorted(zip(model.coef_[0], X_train.columns.values.tolist()))]
for item in coef_list:
    print(item)

job_entrepreneur: -0.9322678224144281
poutcome_failure: -0.7360703558361755
is_loan: -0.6882901627079451
job_blue-collar: -0.5956734567765051
job_unemployed: -0.5897709614775752
is_housing: -0.5855613749463311
is_married: -0.45533425225057506
job_technician: -0.4434429614039016
job_services: -0.41934097407934096
job_self-employed: -0.2776596010239553
is_single: -0.2744267718174345
education_primary: -0.24801126222989317
job_housemaid: -0.17467680632343618
job_management: -0.15326446615040326
education_secondary: -0.14223659400871758
job_admin.: -0.12121238888910131
campaign: -0.08326986863874576
poutcome_other: -0.04754007400341722
month: -0.02740659107463574
age: -0.0031618607032899973
day: -0.0007529413012959553
balance: 2.031345304027948e-06
duration: 0.004053718955778423
previous: 0.01911757379686975
education_tertiary: 0.12860374054901927
job_student: 0.17240984850358923
is_default: 0.4342166153789329
job_retired: 0.4747213803112326
was_contacted: 0.7673862962443289
contact_teleph

We can see from the coefficients that a previous outcome is highly predictive of future outcome, with previous success being highly indicitive of customer subscription. Other important features are whether the custoer was reached by phone, or the type of job the customer has.